src/analysis/souffle/tests/arcs_fact_tests/dynamic_rule_test.dl

```diff
@@ -16,48 +16,100 @@
 // access paths are the result of operation expressions with the isOperatorExpr
 // relation.
 
-.decl mult_op(result: AccessPath, operand1: AccessPath, operand2: AccessPath)
-.decl div_op(result: AccessPath, dividend: AccessPath, divisor: AccessPath)
-
-edge(op1, result), edge(op2, result) :- mult_op(result, op1, op2).
-edge(dividend, result), edge(divisor, result) :- div_op(result, dividend, divisor).
+// .decl mult_op(result: AccessPath, operand1: AccessPath, operand2: AccessPath)
+// .decl div_op(result: AccessPath, dividend: AccessPath, divisor: AccessPath)
 
+// We should not add these edges because the mayHaveIntegrityTag and lacksIntegrityTag
+// rules captures the propagation semantics more precisely. In particular, we should
+// not be using the default propagation rules for `mult_op` and `div_op`.
+//
+// edge(op1, result), edge(op2, result) :- mult_op(result, op1, op2).
+// edge(dividend, result), edge(divisor, result) :- div_op(result, dividend, divisor).
 isAccessPath("a").
 isAccessPath("b").
 isAccessPath("c").
 isAccessPath("d").
 isAccessPath("e").
 
-mult_op("c", "a", "b").
-div_op("e", "c", "d").
+// c = a / b
+// e = c * d
+div_op("c", "a", "b").
+mult_op("e", "c", "d").
 
+// Commenting any of these will cause the expecations to fail.
 hasAppliedIntegrityTag("a", "defaultOwner", "TimestampMillis").
 hasAppliedIntegrityTag("b", "defaultOwner", "Constant_1000").
 hasAppliedIntegrityTag("d", "defaultOwner", "Constant_1000").
 
-// This is the first policy rule.
+// Propagation rules for div_op
 mayHaveIntegrityTag(ap, owner, "TimestampSeconds") :-
   div_op(ap, operand1, operand2),
   mayHaveIntegrityTag(operand1, owner, "TimestampMillis"),
   mayHaveIntegrityTag(operand2, owner, "Constant_1000").
 
+// Default propagation rules.
+mayHaveIntegrityTag(ap, owner, tag) :-
+  div_op(ap, operand1, _),
+  mayHaveIntegrityTag(operand1, owner, tag),
+  tag != "TimestampMillis".
+
+mayHaveIntegrityTag(ap, owner, tag) :-
+  div_op(ap, _, operand2),
+  mayHaveIntegrityTag(operand2, owner, tag),
+  tag != "Constant_1000".
+
 // These rules propagating lacks of integrity tags across rules can only remove
 // integrity tags. Remove them for now to illustrate issue.
-// lacksIntegrityTag(ap, owner, "TimestampSeconds") :-
-//  div_op(ap, operand1, operand2),
-//  (lacksIntegrityTag(operand1, owner, "TimestampMillis");
-//   lacksIntegrityTag(operand2, owner, "Constant_1000")).
+lacksIntegrityTag(ap, owner, "TimestampSeconds") :-
+   div_op(ap, operand1, operand2),
+   (lacksIntegrityTag(operand1, owner, "TimestampMillis");
+    lacksIntegrityTag(operand2, owner, "Constant_1000")).
```

Comment:
I wonder about this propagation of `lacks` across the `PolicyRule` in the case where it does not fire. Imagine that we are creating an analysis that tracks, as in `IntegrityTag`, which integers have `DefinitelyOverflowed`. By default, we believe that any arithmetic operation where both inputs have `DefinitelyOverflowed` has, itself, `DefinitelyOverflowed`. Further, we have a policy rule that applies `DefinitelyOverflowed` to the result of any addition where both operands had the `IntegrityTag` `VeryLargeInteger`. Imagine that that policy rule was automatically converted to the following datalog:

```
mayHaveIntegrityTag(ap, owner, "DefinitelyOverflowed") :-
  mayHaveIntegrityTag(op1, owner, "VeryLargeInteger"),
  mayHaveIntegrityTag(op2, owner, "VeryLargeInteger"),
  add_op(ap, op1, op2).
  
lacksIntegrityTag(ap, owner, "DefinitelyOverflowed") :-
  lacksHaveIntegrityTag(op1, owner, "VeryLargeInteger"),
  lacksHaveIntegrityTag(op2, owner, "VeryLargeInteger"),
  add_op(ap, op1, op2).
```

Now imagine that a `+` operator receives two inputs, each of which is tagged with `DefinitelyOverflowed`. By the default propagation rule, the result should also have `DefinitelyOverflowed`. However, the policy rule indicates that in the case where either of the operands does not have `VeryLargeInteger` on it, the result shall lack the integrity tag `DefinitelyOverflowed`. Therefore, the policy rule will end up *clearing* the `IntegrityTag`!

If the above seems a little hackneyed because the overflow analysis is weird (and I admit, it is weird), imagine that two different policy rules apply the same `IntegrityTag` to the same target `AccessPath` with different `IntegrityTag` preconditions. With the current `lacks` propagation, the preconditions of *both* policy rules must be met for the `IntegrityTag` to be applied rather than the intended behavior of *either* policy rule being able to apply the `IntegrityTag`.

Reply:
The rules might be buggy, but let me try to explain what I was trying. To set context, the approach I am trying is to *group* all the rules for `mayHaveIntegrityTag` and `lacksIntegrity` of a specific  tag + operator combination into a single rule. (This would be analogous to implementing the transfer functions that capture the semantics of an operator during abstract interpretation.)

Given that context, what you want to capture is `ap` is `DefinitelyOverflowed` if _any_ of following conditions hold in a _single_ rule:
   - Both `op1` and `op2` are `DefinitelyOverflowed`, or
   - `op1` is `VeryLargeInteger`, or
   - `op2` is `VeryLargeInteger`

In other words, `ap`  has `DefinitelyOverflowed` if the following holds (bear with `!` for a moment):
```
   (!lacksIntegrity(op1, "DefinitelyOverflowed"), !lacksIntegrity(op2, "DefinitelyOverflowed));
   !lacksIntegrity(op1, "VeryLargeInteger");
   !lacksIntegrity(op2, "VeryLargeInteger") 
```
You would capture `lacksIntegrity` with the negation of above, which would translate to the following:
```
lacksIntegrityTag(ap, owner, "DefinitelyOverflowed") :- 
  (lacksHaveIntegrityTag(op1, owner, "DefinitelyOverflowed");
   lacksHaveIntegrityTag(op2, owner, "DefinitelyOverflowed")), 
   lacksIntegrity(op1, "VeryLargeInteger"), 
   lacksIntegrity(op2, "VeryLargeInteger").
```

That said, I agree that it is not quite straightforward to generate these rules. It can easily lead to loads of bugs.  It is not quite readable also. I am beginning to wonder if it is worth all the trouble.